src/engagement_db_to_analysis/code_imputation_functions.py

```diff
@@ -711,6 +714,92 @@ def _impute_consent_withdrawn(user, column_traced_data_iterable, analysis_datase
+                assert somalia_operator_column_config is None, \
+                    f"Detected multiple coding configurations with " \
+                    f"`analysis_location` {AnalysisLocations.SOMALIA_OPERATOR}"
+                somalia_operator_column_config = coding_config_to_column_config(coding_config, dataset_config.raw_dataset)
+            if coding_config.analysis_location == AnalysisLocations.SOMALIA_ZONE:
+                assert somalia_zone_column_config is None, \
+                    f"Detected multiple coding configurations with " \
+                    f"`analysis_location` {AnalysisLocations.SOMALIA_ZONE}"
+                somalia_zone_column_config = coding_config_to_column_config(coding_config, dataset_config.raw_dataset)
+
+    # If we don't find both, return.
+    if somalia_operator_column_config is None or somalia_zone_column_config is None:
+        log.debug(f"Not imputing Somalia zone from operator because there were no configurations for both "
+                  f"Somalia zone and operator")
+        return
+
+    # Search the column_td for messages with zones that are 'NC' and impute with operator-derived zone codes.
+    log.info(f"Imputing 'NC' Somalia zone labels from operator codes...")
+    imputed_labels = 0
+    for column_td in column_traced_data_iterable:
+        # Get the existing zone labels for this td, and check if they have normal/NC codes
+        zone_labels = column_td[somalia_zone_column_config.coded_field]
+        zone_codes = [somalia_zone_column_config.code_scheme.get_code_with_code_id(l["CodeID"]) for l in zone_labels]
+
+        has_nc_code = False
+        has_normal_code = False
+        for code in zone_codes:
+            if code.control_code == Codes.NOT_CODED:
+                has_nc_code = True
+            if code.code_type == CodeTypes.NORMAL:
+                has_normal_code = True
+
+        # If there is no NC code, then the zone has been usefully labelled as something else e.g. 'nez', 'NR' etc.
+        # Skip this column_td without modifying it.
+        if not has_nc_code:
+            continue
+
+        assert not has_normal_code
+
+        # This column_td has a zone that is labelled as NC only (+ meta codes).
+        # Derive the zone from the operator instead.
+        zone_string = SomaliaLocations.zone_for_operator_code(column_td[somalia_operator_column_config.raw_field])
+        if zone_string == Codes.NOT_CODED:
+            zone_code = somalia_zone_column_config.code_scheme.get_code_with_control_code(Codes.NOT_CODED)
+        else:
+            zone_code = somalia_zone_column_config.code_scheme.get_code_with_match_value(zone_string)
+
+        new_zone_label = CleaningUtils.make_label_from_cleaner_code(
+            somalia_zone_column_config.code_scheme, zone_code, Metadata.get_call_location()
+        )
+
+        # Replace the nc zone label with the new zone label that was created from the operator
```

Comment:
Please confirm with Somali RAs if OPERATOR_TO_ZONE_MAP is still valid or needs to be updated